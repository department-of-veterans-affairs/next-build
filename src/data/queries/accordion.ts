import { QueryFormatter } from 'next-drupal-query'
import { ParagraphAccordion } from '@/types/drupal/paragraph'
import { Accordion } from '@/types/formatted/accordion'

export const formatter: QueryFormatter<ParagraphAccordion, Accordion> = (
  entity: ParagraphAccordion
) => {
  return {
    id: entity.id,
    header: entity.field_header || null,
    html: entity.field_rich_wysiwyg.processed,
  }
}
