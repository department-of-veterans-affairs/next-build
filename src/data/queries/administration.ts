import { FieldAdministration } from '@/types/drupal/field_type'

export interface Administration {
  entityId: number
  name: string | null
}

/**
 * Formats a Drupal administration field into our standardized Administration interface.
 * 
 * In our Drupal content, the `field_administration` field points to a taxonomy Term
 * that is currently called "Section", but "Section" is only a human label added to
 * it in the Drupal interface, and that name has evolved over time. It was originally,
 * "administration" (and thus the name of the field). Since the only machine name of
 * it is `field_administration`, we'll continue to call it "administration" in our code.
 */
export const formatAdministration = (
  administration: FieldAdministration | null | undefined
): Administration => ({
  entityId: administration?.drupal_internal__tid || null, // "t" is for taxonomy term
  name: administration?.name || null,
})
