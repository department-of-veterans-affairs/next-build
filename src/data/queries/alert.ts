// Define the query params for fetching block--alert.
import { BlockAlert } from '@/types/drupal/block'
import { QueryFormatter } from 'next-drupal-query'
import { Alert } from '@/types/formatted/alert'

export const formatter: QueryFormatter<BlockAlert, Alert> = (
  entity: BlockAlert
) => {
  return {
    alertType:
      entity.field_alert_type === 'information'
        ? 'info'
        : entity.field_alert_type,
    id: entity.id,
    title: entity.field_alert_title,
    content: {
      header: entity.field_alert_content.field_text_expander,
      text: entity.field_alert_content.field_wysiwyg.processed,
    },
  }
}
