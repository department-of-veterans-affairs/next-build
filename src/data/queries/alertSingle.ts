import { QueryFormatter } from 'next-drupal-query'
import { ParagraphAlertSingle } from '@/types/drupal/paragraph'
import {
  AlertSingle,
  AlertBlock,
  AlertNonReusable,
} from '@/types/formatted/alert'
import { formatParagraph } from '@/lib/drupal/paragraphs'
import { queries } from '.'

// TODO:
//
// EDIT:
// It seems *maybe* that next-drupal-query is designed to handle this already.
// You can pass an `id` to queries.getParams().
// E.g.
// queries.getParams('paragraph--alert_single'))
//
// We don't need this here:
// export const params: QueryParams<null> = () => {
//   return queries
//     .getParams()
//     .addInclude([
//       'field_alert_block_reference',
//       'field_alert_block_reference.field_alert_content',
//       'field_alert_non_reusable_ref',
//       'field_alert_non_reusable_ref.field_va_paragraphs',
//     ])
// }
//
// ...but it seems like we might want to define
// paragraph-specific `includes` within the paragraph's
// query file (this file), and then somehow reference these in
// the node query when we need to include fields of a specific
// paragraph type.
//
// E.g.
// export const include = [
//   'field_alert_block_reference',
//   'field_alert_block_reference.field_alert_content',
//   'field_alert_non_reusable_ref',
//   'field_alert_non_reusable_ref.field_va_paragraphs',
// ]
//
// /* Some util file*/
// export const getIncludedSubFields = (
//   parentField: string,
//   subFields: string[]
// ) => [parentField, ...subFields.map((subField) => `${parentField}.${subField}`)]
//
// /* src/data/queries/resourcesSupport.ts */
// import { include } from '@/data/queries/alertSingle'
// export const params: QueryParams<null> = () => {
//   return queries.getParams().addInclude([
//     ...getIncludedSubFields('field_alert_single', include)
//   ])
// }

export const formatter: QueryFormatter<ParagraphAlertSingle, AlertSingle> = (
  entity: ParagraphAlertSingle
) => {
  return {
    type: entity.type as AlertSingle['type'],
    id: entity.id,
    entityId: entity.drupal_internal__id,
    alertSelection:
      entity.field_alert_selection as AlertSingle['alertSelection'],
    blockReference: queries.formatData(
      'block--alert',
      entity.field_alert_block_reference
    ) as AlertBlock,
    nonReusableRef: formatParagraph(
      entity.field_alert_non_reusable_ref
    ) as AlertNonReusable,
  }
}
