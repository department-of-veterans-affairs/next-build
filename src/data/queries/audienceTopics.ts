// Define the query params for fetching node--news_story.
import { ParagraphAudienceTopics } from '@/types/drupal/paragraph'
import { QueryFormatter } from 'next-drupal-query'
import { AudienceTopic, AudienceTopics } from '@/types/formatted/audienceTopics'

const getTagsList = (
  entity: ParagraphAudienceTopics
): AudienceTopic[] | null => {
  if (!entity) return null

  const {
    field_topics: fieldTopics = [],
    field_audience_beneficiares: fieldAudienceBeneficiares,
    field_non_beneficiares: fieldNonBeneficiares,
  } = entity

  const topics = fieldTopics.map((topic) => ({
    id: topic.id,
    href: topic.path?.alias,
    name: topic.name,
    categoryLabel: 'Topics',
  }))

  const audiences = [
    ...(fieldAudienceBeneficiares || []),
    ...(fieldNonBeneficiares || []),
  ]
    .filter((tag) => !!tag)
    .map((audience) => ({
      id: audience.id,
      href: audience.path?.alias,
      name: audience.name,
      categoryLabel: 'Audience',
    }))

  const tagList = [...topics, ...audiences]

  return tagList.sort((a, b) => a.categoryLabel.localeCompare(b.categoryLabel))
}

/**
 * This formatter is unusual in that it takes a paragraph entity and digs into
 * its data to return multiple instances of similar values from it.
 */
export const formatter: QueryFormatter<
  ParagraphAudienceTopics,
  AudienceTopics
> = (entity: ParagraphAudienceTopics) => {
  return {
    type: entity.type as AudienceTopics['type'],
    id: entity.id,
    entityId: entity.drupal_internal__id,
    tags: getTagsList(entity),
  }
}
