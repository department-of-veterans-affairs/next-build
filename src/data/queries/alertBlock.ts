// Define the query params for fetching block--alert.
import { BlockAlert } from '@/types/drupal/block'
import { QueryFormatter } from 'next-drupal-query'
import { AlertBlock, AlertType } from '@/types/formatted/alert'

export const formatter: QueryFormatter<BlockAlert, AlertBlock> = (
  entity: BlockAlert
) => {
  if (!entity) {
    return null
  }

  return {
    alertType: (entity.field_alert_type === 'information'
      ? 'info'
      : entity.field_alert_type) as AlertType,
    id: entity.id,
    title: entity.field_alert_title,
    content: {
      header:
        entity.field_alert_content.field_text_expander ??
        entity.field_alert_heading,
      text: entity.field_alert_content?.field_wysiwyg?.processed ?? '',
    },
  }
}
