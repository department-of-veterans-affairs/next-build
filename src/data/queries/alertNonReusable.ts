// Define the query params for fetching block--alert.
import { BlockAlert } from '@/types/drupal/block'
import { QueryFormatter } from 'next-drupal-query'
import { AlertType, AlertNonReusable } from '@/types/formatted/alert'
import { ParagraphNonReusableAlert } from '@/types/drupal/paragraph'
import { formatParagraph } from '@/lib/drupal/paragraphs'

export const formatter: QueryFormatter<
  ParagraphNonReusableAlert,
  AlertNonReusable
> = (entity: ParagraphNonReusableAlert) => {
  if (!entity) {
    return null
  }

  return {
    type: entity.type as AlertNonReusable['type'],
    id: entity.id,
    entityId: entity.drupal_internal__id,
    alertType: (entity.field_alert_type === 'information'
      ? 'info'
      : entity.field_alert_type) as AlertType,
    heading: entity.field_alert_heading,
    paragraphs: entity.field_va_paragraphs.map(formatParagraph),
  }
}
