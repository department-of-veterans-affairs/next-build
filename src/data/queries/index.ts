import { createQueries } from 'next-drupal-query'
import * as NewsStory from './newsStory'
import * as NewsStoryTeaser from './newsStoryTeaser'
import * as StoryListing from './storyListing'
import * as QuestionAnswer from './questionAnswer'
import * as ExpandableText from './expandableText'
import * as LinkTeaser from './linkTeaser'
import * as MediaImage from './mediaImage'
import * as Banner from './banners'
import * as PersonProfile from './personProfile'
import * as Button from './button'
import * as AudienceTopics from './audienceTopics'
import * as Alert from './alert'
import * as EmailContact from './emailContact'
import * as BenefitsHub from './benefitsHub'
import * as Wysiwyg from './wysiwyg'
import * as StaticPathResources from './staticPathResources'
import * as HeaderFooter from './headerFooter'

export const queries = createQueries({
  // standard Drupal entity data queries
  'node--news_story': NewsStory,
  'node--news_story--teaser': NewsStoryTeaser,
  'node--story_listing': StoryListing,
  'node--q_a': QuestionAnswer,
  'node--person_profile': PersonProfile,
  'node--landing_page': BenefitsHub, // "Benefits Hub Landing Page"
  'paragraph--audience_topics': AudienceTopics,
  'paragraph--button': Button,
  'paragraph--email_contact': EmailContact,
  'paragraph--expandable_text': ExpandableText,
  'paragraph--link_teaser': LinkTeaser,
  'paragraph--rich_text_char_limit_1000': Wysiwyg,
  'paragraph--wysiwyg': Wysiwyg,
  'media--image': MediaImage,
  'block--alert': Alert,

  // Custom queries 
  'banner--alerts_lookup': Banner,
  'header-footer-data': HeaderFooter,

  // Static Path Generation
  'static-path-resources': StaticPathResources,
})
