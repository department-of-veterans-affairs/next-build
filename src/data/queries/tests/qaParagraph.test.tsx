import { ParagraphQA } from '@/types/drupal/paragraph'
import { queries } from '@/data/queries'
import mockData from '@/mocks/qaParagraph.mock.json'

const QaParagraphMock: ParagraphQA = mockData

describe('QaParagraph formatData', () => {
  let windowSpy

  beforeEach(() => {
    windowSpy = jest.spyOn(window, 'window', 'get')
  })

  afterEach(() => {
    windowSpy.mockRestore()
  })

  test('outputs formatted data', () => {
    windowSpy.mockImplementation(() => undefined)

    expect(
      queries.formatData('paragraph--q_a', QaParagraphMock)
    ).toMatchSnapshot()
  })
})
