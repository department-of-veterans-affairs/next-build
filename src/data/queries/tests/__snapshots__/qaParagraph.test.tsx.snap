// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QaParagraph formatData outputs formatted data 1`] = `
{
  "answers": [
    {
      "format": "rich_text",
      "processed": "<p>asdfasdf</p>",
      "value": "<p>asdfasdf</p>
",
    },
  ],
  "question": "asdf",
}
`;
